Plugin: Document the project structure and conventions used in the `lib` folder for PHP code (#39603)
WordPress/gutenberg: lib/README.md

# Gutenberg PHP

## File structure

Gutenberg adds features to WordPress Core using PHP hooks and filters. Some
features, once considered stable and useful, are merged into Core during a Core
release. Some features remain in the plugin forever or are removed.

> To make it easier for contribtors to know which features need to be merged to

Reviewer: ```suggestion
To make it easier for contributors to know which features need to be merged to
```

Core and which features can be deleted, Gutenberg uses the following file
structure for its PHP code:

- `lib/experimental` - Experimental features that exist only in the plugin. They
  are not ready to be merged to Core.
- `lib/stable` - Stable features that exist only in the plugin. They could one
  day be merged to Core, but not yet.
- `lib/compat/wordpress-X.Y` - Stable features that are intended to be merged to
  Core in the future X.Y release, or that were previously merged to Core in the
  X.Y release and remain in the plugin for backwards compatibility when running
  the plugin on older versions of WordPress.

## Best practices

### Prefer the `wp` prefix

For features that may be merged to Core, it's best to use a `wp_` prefix for functions or a `WP_` prefix for classes. 

This applies to both experimental and stable features.

> Using the `wp_` prefix facilitates the process of renaming functions and classes from `gutenberg_` to `wp_` if the feature is merged to Core.

Reviewer 2: Does the phrase `facilitates` here mean "makes it easier to", or "avoids us having to"? It sounds like the suggestion here is to use the `wp_` prefix instead of `gutenberg_` where possible, to reduce the number of functions / classes that need to be renamed? (Just wanted to make sure I understand 😀)

Author: Re-worded this. Your understanding is right!

Functions that are intended solely for the plugin, e.g., plugin infrastructure, should use the `gutenberg_` prefix.

> 

Reviewer 3: Any thoughts on naming conventions for when we're updating a function from core?

For example, this change from https://github.com/WordPress/gutenberg/pull/38681/files?w=1#diff-089523f9c993d53f6f02926d28f7a80394d93740e7d16e47e17244547b2b8f5e. The core action is removed and the plugin action is added to replace it.

My intention with that change was that the `gutenberg_` version of the function in the plugin would replace the `wp_` version when porting to core. If we're reserving that prefix for things that should never go into core, we need another convention.

Author: I don't think we really have a good strategy for this kind of stuff. We have a similar nightmare with `WP_Theme_JSON` needing `Gutenberg_Theme_JSON` and `WP_Theme_JSON_5_9` 😅

Using `gutenberg_` seems fine to me, though it will get tricky when the new `gutenberg_` function makes it into WP 6.0 and we want to make more changes in 6.1 We won't be able to tell whether the `wp_` function that exists is the 5.9 version or the 6.0 version.

> #### Good

Reviewer 4: Here we could clarify in what context this example is good.

```suggestion
#### Feature to be merged to Core
```

```php
function wp_get_navigation( $slug ) { ... }
```

> #### Not so good

Reviewer 4: Likewise, we could clarify when this should be used:

```suggestion
#### Plugin infrastructure that will never go in Core
```

```php
function gutenberg_get_navigation( $slug ) { ... }
```

> ### Group PHP code by _feature_

Reviewer 5: I actually think the `lib/compat` folder should group its code by "core" php file. Meaning if the change needs to happen in `edit-form-blocks.php` in 5.9, I'd put it in `lib/compat/5.9/edit-form-blocks.php` and for new files, try to come up with the best file name that will suite core merge. In practice, this is also very close to "group by feature".

The main goal is to simplify the work for the person back porting.

Author: The downside of that though is that you can't move features in and out of "experimental" status e.g. using `git mv lib/experimental/cool-new-feature.php lib/compat/wordpress-6.0/cool-new-feature.php`. Not sure if we do this in practice, mind you 😀

Developers should organize PHP into files or folders by _feature_, not by _component_.

> Also, developers should call `add_action` and `add_filter` immediately after the function being hooked is defined.

Reviewer 2: Totally optional, but I found this sentence a little hard to parse at first. How about:

```suggestion
When defining a function that will be hooked, developers should call `add_action` and `add_filter` immediately after the function declaration.
```

Author: Sure!

These two practices make it easier for PHP code to start in one folder (e.g., `lib/experimental`) and eventually move to another using a simple `git mv`.

#### Good

```php
// lib/experimental/navigation.php

function wp_get_navigation( $slug ) { ... }

function wp_register_navigation_cpt() { ... }

add_action( 'init', 'wp_register_navigation_cpt' );
```

#### Not so good

```php
// lib/experimental/functions.php

function wp_get_navigation( $slug ) { ... }

// lib/experimental/post-types.php

function wp_register_navigation_cpt() { ... }

// lib/experimental/init.php
add_action( 'init', 'wp_register_navigation_cpt' );
```

### Wrap functions and classes with `! function_exists` and `! class_exists`

Developers should take care to not define functions and classes that are already defined.

> When writing new function and classes, it's good practice to use `! function_exists` and `! class_exists`.

Reviewer 2: Tiny grammar fix:

```suggestion
When writing new functions and classes, it's good practice to use `! function_exists` and `! class_exists`.
```

> When writing new function and classes, it's good practice to use `! function_exists` and `! class_exists`.

Reviewer 2: Tiny grammar fix:

```suggestion
When writing new functions and classes, it's good practice to use `! function_exists` and `! class_exists`.
```

If Core has defined a symbol once and then Gutenberg defines it a second time, fatal errors will occur.

Wrapping functions and classes avoids such errors if the feature is merged to Core. 

#### Good

```php
// lib/experimental/navigation/navigation.php

if ( ! function_exists( 'wp_get_navigation' ) ) {
	function wp_get_navigation( $slug ) { ... }
}

// lib/experimental/navigation/class-wp-navigation.php

if ( class_exists( 'WP_Navigation' ) ) {

> 	return;

Reviewer 2: Nice, TIL you can call return at root level in a PHP file (i.e. not inside a function) to skip execution 👍

Author: PHP is pretty lawless.

}

class WP_Navigation { ... }
```

#### Not so good

```php
// lib/experimental/navigation/navigation.php

function wp_get_navigation( $slug ) { ... }

// lib/experimental/navigation/class-gutenberg-navigation.php

class WP_Navigation { ... }
```

Furthermore, a quick codebase search will also help you know if your new method is unique.

### Note how your feature should look when merged to Core

> Developers should write a brief note about how their feature should be merged to Core, for example, which Core file or function should be patched. 

Reviewer 4: This will help so much with backports to Core ⭐

# Gutenberg PHP

## File structure

Gutenberg adds features to WordPress Core using PHP hooks and filters. Some
features, once considered stable and useful, are merged into Core during a Core
release. Some features remain in the plugin forever or are removed.

To make it easier for contribtors to know which features need to be merged to
Core and which features can be deleted, Gutenberg uses the following file
structure for its PHP code:

- `lib/experimental` - Experimental features that exist only in the plugin. They
  are not ready to be merged to Core.
- `lib/stable` - Stable features that exist only in the plugin. They could one
  day be merged to Core, but not yet.

> - `lib/compat/wordpress-X.Y` - Stable features that are intended to be merged to

Reviewer 6: Not suggesting we do it right away, but do  you think it'd be helpful for folks if expanded the compat topic at some stage? For example rationale, preferred naming conventions when extending classes, things to test, and so on. 

There are also plenty of lessons to draw from PRs such as https://github.com/WordPress/gutenberg/pull/38883 and https://github.com/WordPress/gutenberg/pull/38671

Or maybe it's just too fuzzy an area to pin down. 🌪️

Author: Yes absolutely. This should be a living document, like all our docs. I think we're still figuring a lot of that out, so it's hard to pen down right now. If you have any suggestions on things we can add now then please go ahead!
